### PR TITLE
[WIP] streaming http response

### PR DESCRIPTION
**Pull request by bartlomieju**
Bare minimum streaming response. I want to continue on top of work in #15

### CODE REVIEW

File: http.ts [hunk 1]
```diff
@@ -82,7 +82,7 @@ export async function* serve(addr: string) {
   listener.close();
 }
 
-export async function listenAndServe(addr: string, handler: (req: ServerRequest) => void) {
+export async function listenAndServe(addr: string, handler: (ServerRequest) => void) {
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply http.ts]: this shouldn't be changed, thanks for nit

**Comment by bartlomieju** [comment]
@ry PTAL if this is viable

File: http.ts [hunk 2]
```diff
@@ -96,16 +96,23 @@ export async function listenAndServe(
 export interface Response {
   status?: number;
   headers?: Headers;
-  body?: Uint8Array;
+  body?: Uint8Array | Reader;
```
Comment (ry) [inline http.ts]: 👍

File: http.ts [hunk 3]
```diff
@@ -96,16 +96,23 @@ export async function listenAndServe(
...
-    const bodyLength = r.body ? r.body.byteLength : 0;
-    r.headers.append("Content-Length", bodyLength.toString());
+
+  if (r.body) {
+    if (!r.headers.has("content-length")) {
+      if (r.body instanceof Uint8Array) {
+        const bodyLength = r.body.byteLength;
+        r.headers.append("Content-Length", bodyLength.toString());
+      } else {
+        r.headers.append("Transfer-Encoding", "chunked");
```
Comment (ry) [inline http.ts]: 👍

File: http.ts [hunk 4]
```diff
@@ -116,6 +123,29 @@ export class ServerRequest {
   headers: Headers;
   w: BufWriter;
 
+  private async _streamBody(body: Reader, bodyLength: number) {
+    let n = 0;
+    for await (const chunk of toAsyncIterator(body)) {
+      n += await this.w.write(chunk);
+    }
```
Comment (ry) [inline http.ts]: You can use the deno.copy function here.

**Comment by bartlomieju** [comment]
@ry I added two simple test cases. Do you want me to port them 1-to-1? If so, at the moment there are already some differences between implementations like:
- our headers are written all lowercase vs capitalized in Go's impl
- missing `close` option for `Response`

Let me know which direction you prefer

**Comment by ry** [comment]
@bartlomieju This is great! I'll just add a link to the file we're porting. Let's aim to also add the close option eventually - Go's HTTP API is very well thought out - if we follow it we'll be able to move quickly by porting tests.

**Pull request merged**